### PR TITLE
feat: fetch and use the limits from the JWT Token

### LINKED ISSUE #460: Allow the volume configuration to be defined in the auth JWT
Currently the configurations are globally defined.
Is it possible to define those on a request level using data from the JWT?

### PR DESCRIPTION
**Pull request by im-adithya**
Fixes #460

This would enable setting different limits for different users instead of just one global setting.

⚠️ TODO: Add tests

### CODE REVIEW

File: lnd/config.go [hunk 1]
```diff
@@ -22,6 +22,14 @@ type Config struct {
 	LNDClusterPubkeys            string  `envconfig:"LND_CLUSTER_PUBKEYS"` //comma-seperated list of public keys of the cluster
 }
 
+type Limits struct {
```
Comment (kiwiidb) [inline lnd/config.go:25]: These limits are unrelated to LND and thus this struct should not be in the LND package.

File: lib/tokens/jwt.go [hunk 2]
```diff
@@ -15,12 +16,17 @@ import (
...
+	MaxSendVolume      int64 `json:"maxSendVolume"`
+	MaxSendAmount      int64 `json:"maxSendAmount"`
+	MaxReceiveVolume   int64 `json:"maxReceiveVolume"`
+	MaxReceiveAmount   int64 `json:"maxReceiveAmount"`
+	MaxAccountBalance  int64 `json:"maxAccountBalance"`
 	jwt.StandardClaims
 }
 
-func Middleware(secret []byte) echo.MiddlewareFunc {
+func Middleware(secret []byte, limits *lnd.Limits) echo.MiddlewareFunc {
```
Comment (kiwiidb) [inline lib/tokens/jwt.go:29]: I would change nothing to the signature of this function.

File: lnd/config.go (hunk 1 above)
Comment (kiwiidb) [inline lnd/config.go:25]: These limits are unrelated to LND and thus this struct should not be in the LND package.

File: lib/tokens/jwt.go (hunk 2 above)
Comment (kiwiidb) [inline lib/tokens/jwt.go:29]: I would not change the arguments of this function.

File: lib/tokens/jwt.go [hunk 3]
```diff
@@ -38,6 +44,31 @@ func Middleware(secret []byte) echo.MiddlewareFunc {
 		token := c.Get("UserJwt").(*jwt.Token)
 		claims := token.Claims.(*jwtCustomClaims)
 		c.Set("UserID", claims.ID)
+		if limits.MaxSendVolume == 0 {
```
Comment (kiwiidb) [inline lib/tokens/jwt.go:47]: I would just call C.set for everything, no if, no else.

File: lib/service/user.go [hunk 4]
```diff
@@ -279,3 +274,24 @@ func (svc *LndhubService) GetVolumeOverPeriod(ctx context.Context, userId int64,
 	}
 	return result, nil
 }
+
+func (svc *LndhubService) GetLimitsFromContext(c echo.Context) (limits *lnd.Limits) {
```
Comment (kiwiidb) [inline lib/service/user.go:278]: I would create a function defined on `svc *LNDhubService` with the default limits, that are fetched from `svc.Config`. Then I would overwrite these values with the values from the context, if these are larger than 0 (since they will be set always they will never be nil)

File: controllers/addinvoice.ctrl.go [hunk 5]
```diff
@@ -37,6 +37,7 @@ func (controller *AddInvoiceController) AddInvoice(c echo.Context) error {
 
 func AddInvoice(c echo.Context, svc *service.LndhubService, userID int64) error {
 	var body AddInvoiceRequestBody
+	limits := svc.GetLimitsFromContext(c)
```
Comment (kiwiidb) [inline controllers/addinvoice.ctrl.go:40]: There is no need to already do `GetLimitsFromContext` here. We have the context all the way down, we can just fetch the limits when we need them, there is no need to change the arguments of all these functions

**Review by kiwiidb — CHANGES_REQUESTED** [review]
I think the PR can be simplified quite a bit, see comments.

File: lnd/config.go (hunk 1 above)
Comment (im-adithya) [reply lnd/config.go:25]: Shifted to `db/models/user.go`

File: lib/service/user.go (hunk 4 above)
Comment (im-adithya) [reply lib/service/user.go]: Changed to `GetLimits`

**Review by kiwiidb — APPROVED** [review]
